File: electron/handlers/sessions.js
const path = require('path');
const fs = require('fs');
const configService = require('../services/configService');
const sessionParser = require('../services/sessionParser');

/**
 * Validate that a folder path is absolute, exists as a directory,
 * and is in the configured folder list.
 * @param {string} folderPath
 * @param {object} config
 */
function validateFolderAccess(folderPath, config) {
  if (!folderPath || typeof folderPath !== 'string') {
    throw new Error('Folder path is required');
  }

  if (!path.isAbsolute(folderPath)) {
    throw new Error(`Folder path must be absolute: ${folderPath}`);
  }

  let stat;
  try {
    stat = fs.statSync(folderPath);
  } catch (_err) {
    throw new Error(`Folder does not exist or is not accessible: ${folderPath}`);
  }

  if (!stat.isDirectory()) {
    throw new Error(`Folder does not exist or is not a directory: ${folderPath}`);
  }

  if (!config.folders.some((f) => f === folderPath)) {
    throw new Error('Folder not in configured list');
  }
}

function registerSessionHandlers(ipcMain) {
  ipcMain.handle('get-sessions', async (_event, folderPath) => {
    try {
      const config = configService.loadConfig();
      validateFolderAccess(folderPath, config);

      const sessions = await sessionParser.getSessions(folderPath);

      // Sort by timestamp descending (newest first) - immutable sort via spread
      const sorted = [...sessions].sort((a, b) => {
        const ta = new Date(a.timestamp).getTime();
        const tb = new Date(b.timestamp).getTime();
        return tb - ta;
      });

      return sorted;
    } catch (err) {
      throw new Error(`Failed to get sessions: ${err.message}`);
    }
  });

  ipcMain.handle('get-session-detail', async (_event, sessionId, folderPath) => {
    try {
      const config = configService.loadConfig();
      validateFolderAccess(folderPath, config);

      const detail = await sessionParser.getSessionDetail(sessionId, folderPath);
      return detail;
    } catch (err) {
      throw new Error(`Failed to get session detail: ${err.message}`);
    }
  });
}

module.exports = { registerSessionHandlers };

File: src-tauri/src/commands/sessions.rs
use crate::models::session::{SessionDetail, SessionInfo};
use crate::services::{config_service, session_parser::SessionParser};
use std::path::Path;
use tauri::State;

/// Validate that a folder path is absolute and exists as a directory.
fn validate_folder_path(folder_path: &str) -> Result<(), String> {
    let path = Path::new(folder_path);
    if !path.is_absolute() {
        return Err(format!("Folder path must be absolute: {}", folder_path));
    }
    if !path.is_dir() {
        return Err(format!("Folder does not exist or is not a directory: {}", folder_path));
    }
    Ok(())
}

/// Get all sessions for a folder, sorted by time descending
#[tauri::command]
pub async fn get_sessions(
    folder_path: String,
    parser: State<'_, SessionParser>,
) -> Result<Vec<SessionInfo>, String> {
    validate_folder_path(&folder_path)?;

    let config = config_service::load_config().map_err(|e| e.to_string())?;
    if !config.folders.iter().any(|f| f == &folder_path) {
        return Err("Folder not in configured list".to_string());
    }

    let mut sessions = parser
        .get_sessions(&folder_path)
        .map_err(|e| e.to_string())?;
    sessions.sort_by(|a, b| b.timestamp.cmp(&a.timestamp));
    Ok(sessions)
}

/// Get detail for a single session
#[tauri::command]
pub async fn get_session_detail(
    session_id: String,
    folder_path: String,
    parser: State<'_, SessionParser>,
) -> Result<SessionDetail, String> {
    validate_folder_path(&folder_path)?;

    let config = config_service::load_config().map_err(|e| e.to_string())?;
    if !config.folders.iter().any(|f| f == &folder_path) {
        return Err("Folder not in configured list".to_string());
    }

    parser
        .get_session_detail(&session_id, &folder_path)
        .map_err(|e| e.to_string())
}
